rfcs/shared/build-system/package-public-api.md

```diff
@@ -0,0 +1,237 @@
+# RFC: Package public API accessors
+
+---
+
+_@hotell_
+
+## Summary
+
+We wanna control which parts of API surface of our packages are being exposed to public.
+
+## Background
+
+During v9 pre-release phase we ran into various situation where we didn't want to propagate APIs as ready to go, thus we established `_unstable` identifiers naming suffix to mark those as unstable. To complement this approach we also introduced `*/unstable` deep imports for API's that are not STABLE. This works to some extent but fails short as its a manual work that is not checked by automation. This led to scenarios that we unintentionally exposed INTERNAL apis to public which led to Breaking changes which we wanna avoid at all costs as we follow proper semver.
+
+## Problem statement
+
+We expose all APIs of our packages as a PUBLIC from barrel files (`src/index.ts`) to consumers, which introduces problems as we can expose apis that were supposed to be used only internally within our repo or only be available in particular pre-release stage (ALPHA/BETA versions).
+
+## Detailed Design or Proposal
+
+We wanna provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.
+
+Note that our approach will be applied only on **Type Declaration emit level** as we consider TypeScript surface as only source of truth for our package APIs.
+
+> This means that if users use only vanilla JavaScript nothing will stop them from using those API's - as we don't support that kind of usage users take the risk that things may/will break for them unfortunately and also they will not get any support from our side.
+
+### Proposed access modifiers:
+
+#### `@internal`
+
+It indicates that an API item is meant only for usage by other NPM packages from the same repo. Third parties _should never use “internal” APIs._
+
+- tooling will trim the declaration from a public API surface.
+- tolling will trim the declaration from `/unstable` API surface.
+
+**Example:**
+
+```ts
+// @filename package-a/src/index.ts
+
+/**
+ *
+ * @internal
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+
+export declare function hello(): void;
+```
+
+#### `@alpha`
```

Comment:
I'm not quite sure how useful the prerelease stage modifiers will be since it's harder for partners to be aware of these tags, not all IDEs will render tsdoc the same way. For the scenarios that the prerelease modifiers cover, IMO `/unstable` entrypoint + `_unstable` suffix would be more impactful

Reply:
> additional context: if needed we can allow @beta to be shipped from stable and only @alpha from unstable but I'd leave that to team decisions ( IMO we should not do that ).

As `@beta` packages could have breaking changes IMO we should not ship them under stable.

To check my understanding: we have following chain for packages `alpha` -> `beta` -> `rc`. If we follow the same for APIs in comments, we should have also `@rc` comment, right? 👾

Reply:
> does this mean that @fluentui/new-component will only be usable from @fluentui/new-component/unstable while the component is in alpha and beta stages ?

yes that's the idea (from TS api POV _